cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -384,6 +413,52 @@ Runnable deprovision(ProgramRunId programRunId, DatasetContext datasetContext,
     return createDeprovisionTask(provisioningTaskInfo, provisioner, taskCleanup);
   }
 
+  /**
+   * Checks if the given provisioner capabilities satisfies all requirement of a program.
+   *
+   * @param provisionerCapabilities provisioner capabilities
+   * @param requirements a {@link Set} of {@link PluginRequirement}
+   *
+   * @return {@link Set} {@link PluginRequirement} which consists of plugin identifier and the requirements which
+   * are not meet
+   */
+  @VisibleForTesting
+  Set<PluginRequirement> getUnfulfilledRequirements(Set<String> provisionerCapabilities,
+                                                    Set<PluginRequirement> requirements) {
+    // create a map of plugin name to unfulfilled requirement if there are any
+    Set<PluginRequirement> unfulfilledRequirements = new HashSet<>();
+    Set<String> capabilities = provisionerCapabilities.stream().map(String::toLowerCase).collect(Collectors.toSet());
+    for (PluginRequirement pluginRequirement : requirements) {
+      Sets.SetView<String> unfulfilledRequirement = Sets.difference(pluginRequirement.getRequirements(), capabilities);
+      if (!unfulfilledRequirement.isEmpty()) {
+        unfulfilledRequirements.add(new PluginRequirement(pluginRequirement.getPluginId(),
+                                                          unfulfilledRequirement.immutableCopy()));
+      }
+    }
+    return unfulfilledRequirements;
+  }
+
+  /**
+   * Aggregates the given {@link Set} of {@link PluginRequirement} on requirements.
+   *
+   * @param requirements the {@link PluginRequirement} which needs to be aggregated.
+   *
+   * @return a {@link Map} of requirement to {@link Set} of {@link PluginId}
+   */
+  @VisibleForTesting
+  Map<String, Set<PluginId>> groupByRequirement(Set<PluginRequirement> requirements) {
```

Comment:
I think we should remove the PluginId class and have type and name as separate fields in PluginRequirement. This method can be changed to just return a `Map<String, Set<String>>`, as the result is only used in a log message anyway. I think something like 'batchsink:Table' is more readable than 'PluginId{type=batchsink,name=Table}' anyway.

Reply:
I had them as individual field but moved it out as a class to avoid the separator while putting in the map as key but since we never split on it again this approach does look better.